### ISSUE TITLE
Redirect Loop when using a chrome based browser for SAML login

### ISSUE THREAD

**Issue by epeters-mx42** [nextcloud/user_saml#169]
Hello Nextcloud team,

I am trying to set up SAML2 authentication with our IdP solution matrix42 MyWorkspace.
It is working pretty well with Firefox 57 and with Internet Eexplorer 11, so I seem to have done somethin right.
But when using Chrome, Vivaldi or Opera, I get a Redirect Loop instead of a working login.

### Steps to reproduce
1. Configure a SAML authentication to Nextcloud
2. Starting the connection with a Referral from our IdP service
3. User gets a dialogue telling him that the website has been stopped because of a Redirect loop
4. Connect to Nextcloud by entering the URL directly: https://u17srv.local/nextcloud/
5. Nextcloud redirects to our IdP service as configured
6. Authentication is done and user is redirected back to Nextcloud

### Expected behaviour
Log in to Nextcloud should work an any browser when referred from our IdP

### Actual behaviour
I logged and compared the behaviour between step 2 and step 4:
In both cases Nextcloud first does a redirect to https://u17srv.local/nextcloud/index.php/login which itself redirects to https://u17srv.local/nextcloud/index.php/apps/user_saml/saml/login
In Step 2 then comes a redirect back to /nextcloud resulting in a redirect loop which gets interrupted by the browser.
In Step 4 I get the redirect to our IdP service and a successful login

The only difference I found in the GET request was that the browser in step 1 sends a referer to the server. Which seems correct to me because it got directed to nextcloud by our service. So what is wrong with that referer that Nextcloud behaves this way?

### Server configuration

**Operating system**: Ubuntu 17.10

**Web server:** Apache 2 with https

**Database:** MariaDB

**PHP version:** 7.1.11

**Nextcloud version: 12.0.4

**Where did you install Nextcloud from: direct download

**List of activated apps:**

Enabled:
  - activity: 2.5.2
  - bruteforcesettings: 1.0.2
  - comments: 1.2.0
  - dav: 1.3.0
  - federatedfilesharing: 1.2.0
  - federation: 1.2.0
  - files: 1.7.2
  - files_pdfviewer: 1.1.1
  - files_sharing: 1.4.0
  - files_texteditor: 2.4.1
  - files_trashbin: 1.2.0
  - files_versions: 1.5.0
  - files_videoplayer: 1.1.0
  - firstrunwizard: 2.1
  - gallery: 17.0.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.0.0
  - nextcloud_announcements: 1.1
  - notifications: 2.0.0
  - oauth2: 1.0.5
  - password_policy: 1.2.2
  - provisioning_api: 1.2.0
  - serverinfo: 1.2.0
  - sharebymail: 1.2.0
  - survey_client: 1.0.0
  - systemtags: 1.2.0
  - theming: 1.3.0
  - twofactor_backupcodes: 1.1.1
  - updatenotification: 1.2.0
  - user_saml: 1.4.0
  - workflowengine: 1.2.0
Disabled:
  - admin_audit
  - encryption
  - files_external
  - user_external
  - user_ldap

**Nextcloud configuration:**

{
    "system": {
        "instanceid": "ocvznefbucea",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "u17srv.local"
        ],
        "datadirectory": "\/var\/www\/nextcloud\/data",
        "overwrite.cli.url": "https:\/\/u17srv.local\/nextcloud",
        "dbtype": "mysql",
        "version": "12.0.4.3",
        "dbname": "nextcloud",
        "dbhost": "localhost",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true
    }
}

### Client configuration
**Browser: Vivaldi 1.13.1008.36, Chrome 62.0.3202.94, Opera 49.0.2725.47
Also tested: Firefox 57 and Internet Explorer 11 - There the error does not occur

**Operating system: Windows 10 1709

#### Nextcloud log (data/owncloud.log)

there is no log entry from the error

#### Browser log

I logged the whole procedure with fiddler.
Attached is the wohle log export.
[SAML_Nextcloud_ErrorwhenReferrer.zip](https://github.com/nextcloud/user_saml/files/1563079/SAML_Nextcloud_ErrorwhenReferrer.zip)

GET- Request from step 2:
(entry 29 from my log export)

GET https://u17srv.local/nextcloud/index.php/apps/user_saml/saml/login?requesttoken=yZ/JNcnCHAMIASqO/G3SpwuMwSw/LQ2C7bzWq%2BmSdb0%3D%3Ar/Xievzya1dnYGzIjlXl4Dz4uUR4d0LM29mc/r7BHY4%3D&originalUrl=https%3A//u17srv.local/nextcloud/index.php/apps/files HTTP/1.1
Host: u17srv.local
Connection: keep-alive
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.97 Safari/537.36 Vivaldi/1.94.1008.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Referer: https://accounts-api.matrix42.com/launcher/7641b19d-b52a-44a2-959f-889c3a1bf86a
Accept-Encoding: gzip, deflate, br
Accept-Language: de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7
Cookie: oc_sessionPassphrase=ZlK79en94E1S%2BltO48Iizyg%2FgHGeT0zeE9cnryhyhjczg%2Fj9oo039zOe%2FZikQRNx35FKhHD7k4L%2BupePCBj4cFVIGmVEuY4IywmYTKLRk9mpHZx9cTedGsz9NgpLKjYt; ocvznefbucea=n2a2jh2qtn5ulg89hbcr6hotr4; nc_sameSiteCookielax=true

HTTP/1.1 303 See Other
Date: Fri, 15 Dec 2017 09:37:28 GMT
Server: Apache/2.4.27 (Ubuntu)
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-cache, no-store, must-revalidate
Pragma: no-cache
Content-Security-Policy: default-src 'none';base-uri 'none';manifest-src 'self';script-src 'self' 'unsafe-eval';style-src 'self' 'unsafe-inline';img-src 'self' data: blob:;font-src 'self';connect-src 'self';media-src 'self'
X-Frame-Options: SAMEORIGIN
Location: /nextcloud
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
X-Robots-Tag: none
X-Download-Options: noopen
X-Permitted-Cross-Domain-Policies: none
Content-Length: 0
Keep-Alive: timeout=5, max=98
Connection: Keep-Alive
Content-Type: text/html; charset=UTF-8

GET- Request from step 4:
(entry 71 from my log export)

GET https://u17srv.local/nextcloud/index.php/apps/user_saml/saml/login?requesttoken=7Vaca12xYg%2B5IWuKnKSRDpbg/bNTHAHcOz6Pj0xbZ%2BA%3D%3Aizy3JGiBFVvWQC3M7pymSaGUhdsURk6SDVvF2hsID9M%3D&originalUrl=https%3A//u17srv.local/nextcloud/index.php/apps/files HTTP/1.1
Host: u17srv.local
Connection: keep-alive
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.97 Safari/537.36 Vivaldi/1.94.1008.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Accept-Encoding: gzip, deflate, br
Accept-Language: de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7
Cookie: oc_sessionPassphrase=ZlK79en94E1S%2BltO48Iizyg%2FgHGeT0zeE9cnryhyhjczg%2Fj9oo039zOe%2FZikQRNx35FKhHD7k4L%2BupePCBj4cFVIGmVEuY4IywmYTKLRk9mpHZx9cTedGsz9NgpLKjYt; ocvznefbucea=n2a2jh2qtn5ulg89hbcr6hotr4; nc_sameSiteCookielax=true; nc_sameSiteCookiestrict=true

HTTP/1.1 303 See Other
Date: Fri, 15 Dec 2017 09:37:33 GMT
Server: Apache/2.4.27 (Ubuntu)
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-cache, no-store, must-revalidate
Pragma: no-cache
Content-Security-Policy: default-src 'none';base-uri 'none';manifest-src 'self';script-src 'self' 'unsafe-eval';style-src 'self' 'unsafe-inline';img-src 'self' data: blob:;font-src 'self';connect-src 'self';media-src 'self'
X-Frame-Options: SAMEORIGIN
Location: https://accounts.matrix42.com/issue/7641b19d-b52a-44a2-959f-889c3a1bf86a/saml2/sso?SAMLRequest=pZJPj9MwEMXv%2FRSr3BPHSdokVlupbPlTqbTVtnDgspo4DrWU2MFjL%2BXb4yTAAhJ7wTeP5%2F1m3pOXCF3bs42zV%2FUgvjiBdnbnz61rFbLxcRU4o5gGlMgUdAKZ5ey8eb9nSRSz3miruW6Dv2QvqwBRGCu1mmS77So4Hl7vj293h8d6QYs4z5pFLjgv0oQCTWJexZDRtGjoIstrXok4n6QfhUHPWQUeG8wmGqITO4UWlPX1mOYhTUI6v8QlS3OWpp8m6dablQrsKL9a2yMjBDjXTlmMOrBG3rIk4rojckCSfJHRipZ1WM0TCLMMkrCcl01YFCVPgVZNsQAymE8Iop6GnH7k80qqWqrPL8dSTU3I3l0up%2FB0PF8myOZnXPdaoeuEOQvzJLn48LB%2F3tzRHM1T1GoOLVHiZnmrXU08Utyi%2FtoT6HskzpMehx3HRb1dDNbjjOVwZ2N2Zv1fzE5YqMHCkvyOfB7Ss4N3vtuedCv5t7E%2BnDfa%2BMz%2FHRCN6FiRddiMrcwp7AWXjRR18AuzaVv99d4IsGIVWONEcEfWs9m0zJ8%2Fff0d&RelayState=https%3A%2F%2Fu17srv.local%2Fnextcloud%2Findex.php%2Fapps%2Fuser_saml%2Fsaml%2Flogin&SigAlg=http%3A%2F%2Fwww.w3.org%2F2000%2F09%2Fxmldsig%23rsa-sha1&Signature=Q9L%2Fe1y8Qj9ntkdkDC%2B9Q%2BwOFsOMIErtOFBLE7KCC5EVnIM5KZIBSRUEnZk0DoXd6f%2F2sHJ6Q%2BuL1M0uI7FfcKjjEmQOsmKzdvNigCNjOrlFWgdmBRzZcc1He4nl6ErzdCUg2IlYC0aA6FCr0hwzqDZH8%2B0taMe2w6eHGANSw11naYHSJXVVJTddUkjO2Sn4r5xaixBD6wekkjMz93HT4R8e58MKu6uya7GU7kuMvPdttPiaiOlVCiNWiyJFzBG%2FWb2hpT4nYnexu818iYxw6VmYXXbIkhc6NP%2B5hxG%2BikrLMZWWscWkjaQBQcFN6lJnJL2fD1n%2FErBowo%2Fc5bn5lA%3D%3D
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
X-Robots-Tag: none
X-Download-Options: noopen
X-Permitted-Cross-Domain-Policies: none
Content-Length: 0
Keep-Alive: timeout=5, max=61
Connection: Keep-Alive
Content-Type: text/html; charset=UTF-8

**Comment by linuxrrze** [comment]
We had/have the same problem here.
Try this pull request - works for us:

https://github.com/nextcloud/user_saml/pull/147

I'm still hoping to get it merged some day.

**Comment by epeters-mx42** [comment]
Hello linuxrrze,
thank you for the solution.

I applied it in my environment and it works here too.

**Comment by linuxrrze** [comment]
Hi epeters-mx42,

Great news!
Let's hope that someone will someday merge that one-liner ;-)

**Comment by linuxrrze** [comment]
As there seems to be some merge activity lately I hope this simple patch finally makes it into the next release!
 So +1 for merge!

**Comment by coredump4** [comment]
This problem still occurs with Chrome 70, and the 1-line patch still works around it as well.

**Comment by amartanius** [comment]
I've fixed this issue after 5 days of pinpointing the source of the issue, by setting the cookies age (expiration) the same as the session age. It seems like the cookie that has the ID (for me is was JSESSIONID since i was working with tomcat)

For my situation in tomcat I've set this:

<session-config>
    <session-timeout>60</session-timeout>
    <cookie-config>
        <http-only>true</http-only>
        <secure>true</secure>
        <max-age>3600</max-age>
    </cookie-config>
    <tracking-mode>COOKIE</tracking-mode>
</session-config>
Now after the session end, the cookie will be deleted too. Making it possible to authenticate again with a new cookie.

Hope this helps

**Comment by amartanius** [comment]
I've fixed this issue after 5 days of pinpointing the source of the issue, by setting the cookies age (expiration) the same as the session age. It seems like the cookie that has the ID (for me is was JSESSIONID since i was working with tomcat)

For my situation in tomcat I've set this:
```

<session-config>
    <session-timeout>60</session-timeout>
    <cookie-config>
        <http-only>true</http-only>
        <secure>true</secure>
        <max-age>3600</max-age>
    </cookie-config>
    <tracking-mode>COOKIE</tracking-mode>
</session-config>
```
Now after the session end, the cookie will be deleted too. Making it possible to authenticate again with a new cookie.
